vortex-array/src/arrays/listview/reset.rs

```diff
@@ -0,0 +1,144 @@
+            let mut max_end = 0usize;
+            for i in 0..self.len() {
+                let offset = self.offset_at(i);
+                let size = self.size_at(i);
+                let end = offset + size;
+                max_end = max_end.max(end);
+            }
+
+            // Slice the elements array to include only the referenced elements.
+            let new_elements = self.elements.slice(min_offset..max_end);
+
+            // Adjust offsets by subtracting the minimum offset (which is the first offset).
+            let first_offset_scalar = self.offsets.scalar_at(0);
+            let new_offsets = compute::sub_scalar(&self.offsets, first_offset_scalar)?;
+
+            return Self::try_new(
+                new_elements,
+                new_offsets,
+                self.sizes.clone(),
+                self.validity.clone(),
+            );
+        }
+
+        // Slow path: Rebuild the elements array for out-of-order offsets.
+        // Since `ListViewArray` allows out-of-order offsets, we need to rebuild the `elements`
+        // array by collecting only the referenced elements in order.
+        let mut elements_builder =
+            builder_with_capacity(self.elements.dtype(), self.elements.len());
+
+        // Use nested matches to dispatch to the generic function with concrete types.
+        let (new_offsets, new_sizes) =
+            match_each_integer_ptype!(self.offsets.dtype().as_ptype(), |O| {
+                match_each_integer_ptype!(self.sizes.dtype().as_ptype(), |S| {
+                    build_offset_size_arrays::<O, S>(self, &mut *elements_builder)?
+                })
+            });
+
+        // Finish the elements builder.
+        let new_elements = elements_builder.finish();
+
+        Self::try_new(new_elements, new_offsets, new_sizes, self.validity.clone())
+    }
+}
+
+/// Generic helper function to build offset/size arrays with any integer type during reset_offsets.
+fn build_offset_size_arrays<O, S>(
+    listview: &ListViewArray,
+    elements_builder: &mut dyn ArrayBuilder,
+) -> VortexResult<(ArrayRef, ArrayRef)>
+where
+    O: OffsetPType,
+    S: OffsetPType,
+{
+    let mut offsets_builder =
+        PrimitiveBuilder::<O>::with_capacity(Nullability::NonNullable, listview.len());
+    let mut sizes_builder =
+        PrimitiveBuilder::<S>::with_capacity(Nullability::NonNullable, listview.len());
+
+    // Create uninit ranges for direct memory access.
+    let mut offsets_range = offsets_builder.uninit_range(listview.len());
+    let mut sizes_range = sizes_builder.uninit_range(listview.len());
+
+    let mut current_offset = 0usize;
+
+    for i in 0..listview.len() {
+        let offset = listview.offset_at(i);
+        let size = listview.size_at(i);
+
+        // Append the elements for this list to the builder.
+        for j in 0..size {
+            elements_builder.append_scalar(&listview.elements.scalar_at(offset + j))?;
```

Comment:
I need to add api to remove usage of scalar at here

Reply:
Builders aren't type erased though. I have an idea that I need to prototype. I think in practice you always know the concrete type when you use it